pymc_marketing/customer_choice/nested_logit.py

```diff
@@ -0,0 +1,783 @@
+                    else:
+                        idxs = [product_indices[i] for i in nest_dict[k][j]]
+                        mid_level[k + "_" + j] = idxs
+                        collected_idxs = collected_idxs + idxs
+                top_level[k] = np.sort(collected_idxs)
+            else:
+                top_level[k] = np.sort([product_indices[i] for i in nest_dict[k]])
+
+        if not mid_level:
+            mid_level = None
+        return top_level, mid_level
+
+    @staticmethod
+    def _prepare_coords(df, alternatives, covariates, f_covariates, nests):
+        """Prepare coordinates for PyMC nested logit model."""
+        if isinstance(f_covariates, np.ndarray) & (f_covariates.size > 0):
+            f_cov = [s.strip() for s in f_covariates[0].split("+")]
+        else:
+            f_cov = []
+        coords = {
+            "alts": alternatives,
+            "alts_probs": alternatives[:-1],
+            "alt_covariates": covariates,
+            "fixed_covariates": f_cov,
+            "nests": nests,
+            "obs": range(len(df)),
+        }
+        return coords
+
+    def preprocess_model_data(
+        self, choice_df, utility_equations
+    ) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+        """Pre-process the model initiation inputs into a format that can be used by the PyMC model."""
+        X, F, alternatives, fixed_covar = self.prepare_X_matrix(
+            choice_df, utility_equations, self.depvar
+        )
+        self.X, self.F, self.alternatives, self.fixed_covar = (
+            X,
+            F,
+            alternatives,
+            fixed_covar,
+        )
+        y, prod_mapping = self._prepare_y_outcome(
+            choice_df, self.alternatives, self.depvar
+        )
+        self.y = y
+        self.prod_indices = prod_mapping
+        top_level, mid_level = self._parse_nesting(
+            self.nesting_structure, self.prod_indices
+        )
+        if mid_level:
+            all_nests = list(top_level.keys()) + list(mid_level.keys())
+            self.nest_indices = {"top": top_level, "mid": mid_level}
+        else:
+            all_nests = list(top_level.keys())
+            self.nest_indices = {"top": top_level}
+
+        self.all_nests = all_nests
+        self.lambda_lkup = dict(zip(all_nests, range(len(all_nests)), strict=False))
+
+        # note: type hints for coords required for mypy to not get confused
+        self.coords: dict[str, list[str]] = self._prepare_coords(
+            choice_df,
+            self.alternatives,
+            self.covariates,
+            self.fixed_covar,
+            self.all_nests,
+        )
+
+        return X, F, y
+
+    def build_model(self, X, y, **kwargs):
+        """Do not use, required by parent class. Prefer make_model()."""
+        return super().build_model(X, y, **kwargs)
+
+    def make_exp_nest(self, U, W, betas_fixed_, lambdas_nests, nest, level="top"):
+        """Calculate within nest probabilities.
```

Comment:
Can you explain a bit more the logic of this in the doctrings?

Reply:
done